pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,375 @@
+	// is set to false.
+	ParamNodeName = "node-name"
+	// ParamResyncPeriod is the resync period for the pod cache as a Duration.
+	ParamResyncPeriod = "resync-period"
+	// ParamUserAgent is the user agent used when talking to the k8s API.
+	ParamUserAgent = "user-agent"
+	// ParamWatchCluster is true if we should watch pods in the entire cluster, false if we should watch pods on our
+	// own node.
+	ParamWatchCluster = "watch-cluster"
+
+	// DefaultAPIQPS is the default maximum amount of queries per second we allow to the Kubernetes API server.
+	DefaultAPIQPS = 5
+	// DefaultAPIQPSBurst is the default amount of queries per second we can burst above the maximum QPS.
+	DefaultAPIQPSBurst = 10
+	// DefaultKubeconfigContext is the default context to use inside the kubeconfig file. "" means use the current
+	// context without switching.
+	DefaultKubeconfigContext = ""
+	// DefaultKubeconfigPath is the default path to the kubeconfig file. "" means use in-cluster auth.
+	DefaultKubeconfigPath = ""
+	// DefaultLabelTagRegex is the default ParamLabelTagRegex. Every label is ignored by default.
+	DefaultLabelTagRegex = ""
+	// DefaultNodeName is the default node name to watch pods on when ParamWatchCluster is false. Defaults to unset
+	// as this will fail fast and alert users to set this appropriately.
+	DefaultNodeName = ""
+	// DefaultResyncPeriod is the default resync period for the pod cache.
+	DefaultResyncPeriod = 5 * time.Minute
+	// DefaultUserAgent is the default user agent used when talking to the k8s API.
+	DefaultUserAgent = "gostatsd"
+	// DefaultWatchCluster is the default watch mode for pods. By default we watch the entire cluster.
+	DefaultWatchCluster = true
+)
+
+// PodInformerOptions represent options for a pod informer.
+type PodInformerOptions struct {
+	ResyncPeriod time.Duration
+	WatchCluster bool
+	NodeName     string
+}
+
+// Provider represents a k8s provider.
+type Provider struct {
+	logger logrus.FieldLogger
+
+	podsInf         cache.SharedIndexInformer
+	factory         informers.SharedInformerFactory
+	annotationRegex *regexp.Regexp
+	labelRegex      *regexp.Regexp
+}
+
+func (p *Provider) EstimatedTags() int {
+	// There is no real way to estimate this for k8s provider as any pod can have arbitrary labels/annotations
+	return 0
+}
+
+// Instance returns pod details from k8s API server watches.
+// ip -> nil pointer if pod was not found.
+// map is returned even in case of errors because it may contain partial data.
+// An "instance", as far as k8s cloud provider is concerned, is an individual pod running in the cluster
+func (p *Provider) Instance(ctx context.Context, IP ...gostatsd.IP) (map[gostatsd.IP]*gostatsd.Instance, error) {
+	instanceIPs := make(map[gostatsd.IP]*gostatsd.Instance, len(IP))
+	var returnErr error
+
+	p.logger.WithField("cacheKeys", p.podsInf.GetIndexer().ListKeys()).Debug("pods in cache")
```

Comment:
`ListKeys()` is always evaluated, is this an expensive call?

Reply:
Yeah it probably isn't the cheapest. I will fix this.

Reply:
It was useful for debugging but I've removed it completely since I couldn't think of a nicer way.